### PR TITLE
Add CodeDomComponentSerializationService tests and fix some bugs

### PR DESCRIPTION
**Pull request by hughbe**
###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/1439)

### CODE REVIEW

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomComponentSerializationService.cs [hunk 1]
```diff
@@ -395,7 +395,7 @@ public override void Close()
                 {
                     Hashtable state = new Hashtable(_objects.Count);
                     DesignerSerializationManager manager = new DesignerSerializationManager(new LocalServices(this, _provider));
-                    if (_provider.GetService(typeof(IDesignerSerializationManager)) is DesignerSerializationManager hostManager)
+                    if (_provider?.GetService(typeof(IDesignerSerializationManager)) is DesignerSerializationManager hostManager)
```
Comment (hughbe) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomComponentSerializationService.cs]: Can construct `new CodeDomComponentSerializationService()` so this used to throw NRE

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomComponentSerializationService.cs [hunk 2]
```diff
@@ -672,16 +672,15 @@ internal static void TraceCode(string name, object code)
             /// </summary>
             void ISerializable.GetObjectData(SerializationInfo info, StreamingContext context)
             {
-                Hashtable resources = null;
-                if (_resources != null)
+                if (info == null)
```
Comment (hughbe) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomComponentSerializationService.cs]: used to throw NRE, now ANE

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomComponentSerializationService.cs [hunk 3]
```diff
@@ -1290,7 +1289,6 @@ internal void Serialize(IDesignerSerializationManager manager, IDictionary objec
                             ArrayList defaultPropList = null;
                             List<string> defaultEventList = null;
                             IEventBindingService ebs = manager.GetService(typeof(IEventBindingService)) as IEventBindingService;
-                            Debug.Assert(ebs != null, "No IEventBindingService is available.  Events will not be serialized correctly");
```
Comment (hughbe) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomComponentSerializationService.cs]: Invalid debug assert - we handle cases where we have a null event binding service

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomComponentSerializationService.cs [hunk 4]
```diff
@@ -1384,7 +1382,6 @@ internal void Serialize(IDesignerSerializationManager manager, IDictionary objec
                     }
                     finally
                     {
-                        Debug.Assert(manager.Context.Current == statementCxt, "Context stack corrupted");
```
Comment (hughbe) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/CodeDomComponentSerializationService.cs]: Invalid if we throw an exception in the middle of the serialization

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/ComponentCache.cs [hunk 5]
```diff
@@ -19,8 +19,7 @@ internal class ComponentCache : IDisposable
         internal ComponentCache(IDesignerSerializationManager manager)
         {
             _serManager = manager;
-            IComponentChangeService cs = (IComponentChangeService)manager.GetService(typeof(IComponentChangeService));
-            if (cs != null)
+            if (manager.GetService(typeof(IComponentChangeService)) is IComponentChangeService cs)
```
Comment (hughbe) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/ComponentCache.cs]: #817

File: src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/ComponentCodeDomSerializer.cs [hunk 6]
```diff
@@ -432,8 +430,7 @@ public override object Serialize(IDesignerSerializationManager manager, object v
                                     // new entry object even if there is still an existing one that is just invalid, and it
                                     // might have dependencies that will be lost.
                                     // we need to make sure we copy over any dependencies that are also tracked.
-                                    ComponentCache.Entry oldEntry = null;
-                                    oldEntry = cache.GetEntryAll(value);
+                                    ComponentCache.Entry oldEntry = cache?.GetEntryAll(value);
```
Comment (hughbe) [inline src/System.Windows.Forms.Design/src/System/ComponentModel/Design/Serialization/ComponentCodeDomSerializer.cs]: If we construct a default service, this throws NRE as IServiceProvider.GetService(typeof(ComponentCache)) returns null

File: src/System.Windows.Forms.Design/tests/UnitTests/CodeDomHelpers.cs [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                }
+            }
+
+            string GetString(string s)
+            {
+                if (s == null)
+                {
+                    return "null";
+                }
+                else if (s.Length == 0)
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/tests/UnitTests/CodeDomHelpers.cs]: redundant `else`s

File: src/System.Windows.Forms.Design/tests/UnitTests/CodeDomHelpers.cs [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                {
+                    return "string.Empty";
+                }
+                else
+                {
+                    return $"\"{s}\"";
+                }
+            }
+            string GetType(CodeTypeReference reference)
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/tests/UnitTests/CodeDomHelpers.cs]: [nit] missing blank line

File: src/System.Windows.Forms.Design/tests/UnitTests/CodeDomHelpers.cs [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+        {
+            if (o is CodeStatement cs)
+            {
+                if (cs.StartDirectives.Count != 0 || cs.EndDirectives.Count != 0)
+                {
+                    throw new NotImplementedException("Directives not supported.");
+                }
+            }
+
+            string GetString(string s)
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/tests/UnitTests/CodeDomHelpers.cs]: Please move local functions to the end of the containing method after the last return statement.

File: src/System.Windows.Forms.Design/tests/UnitTests/CodeDomHelpers.cs [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+            if (o is CodeVariableDeclarationStatement v)
+            {
+                if (v.InitExpression == null)
+                {
+                    return $"new CodeVariableDeclarationStatement({GetType(v.Type)}, {GetString(v.Name)});";
+                }
+
+                return $"new CodeVariableDeclarationStatement({GetType(v.Type)}, {GetString(v.Name)}, {GetConstructionString(v.InitExpression)});";
+            }
+            else if (o is CodeAssignStatement cas)
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/tests/UnitTests/CodeDomHelpers.cs]: redundant `else`s

File: src/System.Windows.Forms.Design/tests/UnitTests/CodeDomHelpers.cs [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+                Type result = Type.GetType(reference.BaseType);
+                if (result != null)
+                {
+                    return $"typeof({result.Name})";
+                }
+
+                return GetString(reference.BaseType);
+            }
+
+            if (o is CodeVariableDeclarationStatement v)
```
Comment (RussKie) [inline src/System.Windows.Forms.Design/tests/UnitTests/CodeDomHelpers.cs]: Can we use switch pattern matching here?
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/switch#type-pattern

**Pull request merged**